unic/char/property/src/macros.rs

```diff
@@ -108,62 +147,220 @@ macro_rules! char_property {
-                ];
+                const VALUES: &[$name] = &[ $($name::$variant,)+ ];
                 VALUES
             }
-
             fn abbr_name(&self) -> &'static str {
-                match *self {
-                    $( $name::$variant => stringify!($abbr), )*
-                }
+                match *self { $( $name::$variant => stringify!($abbr), )* }
             }
-
             fn long_name(&self) -> &'static str {
-                match *self {
-                    $( $name::$variant => stringify!($long), )*
-                }
+                match *self { $( $name::$variant => stringify!($long), )* }
             }
-
             fn human_name(&self) -> &'static str {
-                match *self {
-                    $( $name::$variant => $human, )*
+                match *self { $( $name::$variant => $human, )* }
+            }
+        }
+    };
+
+    // == Binary Property == //
+
+    (
+        $(#[$name_meta:meta])*
+        pub struct $name:ident(bool) {
+            abbr => $abbr_name:expr;
+            long => $long_name:expr;
+            human => $human_name:expr;
+
+            data_table_path => $data_path:expr;
+
+            pub fn $bool_name:ident(&self) -> bool;
+        }
+
+        pub fn $shorthand_name:ident(char) -> bool;
+
+        $(#[$char_trait_meta:meta])*
+        pub trait $char_trait_name:ident for char {
+            $(#[$char_trait_fn_meta:meta])*
+            pub fn $char_trait_fn_name:ident(self) -> bool;
+        }
+
+        $(#[$str_trait_meta:meta])*
+        pub trait $str_trait_name:ident for str {
+            $(#[$str_trait_fn_meta:meta])*
+            pub fn $str_trait_fn_name:ident(&self) -> bool;
+        }
+
+    ) => {
+        $(#[$name_meta])*
+        #[derive(Copy, Clone, Debug, Default, Eq, PartialEq, Hash)]
+        pub struct $name(bool);
+
+        mod data {
+            use super::unic_utils::CharDataTable;
+            pub const TABLE: CharDataTable<()> = include!($data_path);
+        }
+
+        /// Get boolean property value of the character.
+        pub fn $shorthand_name(ch: char) -> bool {
+            $name::of(ch).into()
+        }
+
+        impl $name {
+            /// Get (struct) property value of the character.
+            pub fn of(ch: char) -> Self { $name(data::TABLE.contains(ch)) }
+
+            /// Get boolean property value of the character.
+            pub fn $bool_name(&self) -> bool { self.0 }
+        }
+
+        char_property! {
+            __impl CharProperty for $name;
+            $abbr_name;
+            $long_name;
+            $human_name;
+        }
+
+        impl $crate::TotalCharProperty for $name {
+            fn of(ch: char) -> Self { Self::of(ch) }
+        }
+
+        impl $crate::BinaryCharProperty for $name {
+            fn bool(&self) -> bool { self.$bool_name() }
+        }
+
+        impl From<bool> for $name {
+            fn from(b: bool) -> Self { $name(b) }
+        }
+
+        impl From<$name> for bool {
```

Comment:
Similarly to above, do we actually want this? We already have a fn to turn a property into a bool anyway. Is there any use case for using `Into<bool>`? Otherwise I feel that this impl is actually just bloat rather than useful. It could be added later if we find a use for it.